File: src/EditorFeatures/CSharp/DocumentationComments/XmlTagCompletionCommandHandler.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.ComponentModel.Composition;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.DocumentationComments;
using Microsoft.CodeAnalysis.Host.Mef;
using Microsoft.VisualStudio.Commanding;
using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.Utilities;

namespace Microsoft.CodeAnalysis.Editor.CSharp.DocumentationComments;

[Export(typeof(ICommandHandler))]
[ContentType(ContentTypeNames.CSharpContentType)]
[Name(nameof(XmlTagCompletionCommandHandler))]
[Order(Before = PredefinedCompletionNames.CompletionCommandHandler)]
[method: ImportingConstructor]
[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
internal sealed class XmlTagCompletionCommandHandler(ITextUndoHistoryRegistry undoHistory) : AbstractXmlTagCompletionCommandHandler<
    XmlNameSyntax,
    XmlTextSyntax,
    XmlElementSyntax,
    XmlElementStartTagSyntax,
    XmlElementEndTagSyntax,
    DocumentationCommentTriviaSyntax>(undoHistory)
{
    protected override XmlElementStartTagSyntax GetStartTag(XmlElementSyntax xmlElement)
        => xmlElement.StartTag;

    protected override XmlElementEndTagSyntax GetEndTag(XmlElementSyntax xmlElement)
        => xmlElement.EndTag;

    protected override XmlNameSyntax GetName(XmlElementEndTagSyntax endTag)
        => endTag.Name;

    protected override XmlNameSyntax GetName(XmlElementStartTagSyntax startTag)
        => startTag.Name;

    protected override SyntaxToken GetLocalName(XmlNameSyntax name)
        => name.LocalName;
}

File: src/EditorFeatures/VisualBasic/DocumentationComments/XmlTagCompletionCommandHandler.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.ComponentModel.Composition
Imports System.Diagnostics.CodeAnalysis
Imports Microsoft.CodeAnalysis.DocumentationComments
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Microsoft.VisualStudio.Commanding
Imports Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion
Imports Microsoft.VisualStudio.Text.Operations
Imports Microsoft.VisualStudio.Utilities

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.DocumentationComments
    <Export(GetType(ICommandHandler))>
    <ContentType(ContentTypeNames.VisualBasicContentType)>
    <Name("XmlTagCompletionCommandHandler")>
    <Order(Before:=PredefinedCompletionNames.CompletionCommandHandler)>
    Friend NotInheritable Class XmlTagCompletionCommandHandler
        Inherits AbstractXmlTagCompletionCommandHandler(Of
            XmlNameSyntax,
            XmlTextSyntax,
            XmlElementSyntax,
            XmlElementStartTagSyntax,
            XmlElementEndTagSyntax,
            DocumentationCommentTriviaSyntax)

        <ImportingConstructor>
        <SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification:="Used in test code: https://github.com/dotnet/roslyn/issues/42814")>
        Public Sub New(undoHistory As ITextUndoHistoryRegistry)
            MyBase.New(undoHistory)
        End Sub

        Protected Overrides Function GetStartTag(xmlElement As XmlElementSyntax) As XmlElementStartTagSyntax
            Return xmlElement.StartTag
        End Function

        Protected Overrides Function GetEndTag(xmlElement As XmlElementSyntax) As XmlElementEndTagSyntax
            Return xmlElement.EndTag
        End Function

        Protected Overrides Function GetName(startTag As XmlElementStartTagSyntax) As XmlNameSyntax
            Return DirectCast(startTag.Name, XmlNameSyntax)
        End Function

        Protected Overrides Function GetName(endTag As XmlElementEndTagSyntax) As XmlNameSyntax
            Return endTag.Name
        End Function

        Protected Overrides Function GetLocalName(name As XmlNameSyntax) As SyntaxToken
            Return If(name Is Nothing, Nothing, name.LocalName)
        End Function
    End Class
End Namespace
